### PR TITLE
Enable distributed queries of gowarcserver processes

### LINKED ISSUE #4: Allow distribution of gowarcservers with a "parent->child" relationship
Based on meeting with @maeb. He had an idea of a potential direction to improve gowarcserver.

**Is your feature request related to a problem? Please describe.**
This will solve two problems. 

1. In the loke GUI you can see all the different collections on the main page. If you have a series of collections, then it can be cumbersome to find a given warc record as you have to be aware which collection has the record or manually search each collection
2. Optimize gowarcserver by distributing indexing and searching

**Describe the solution you'd like**

Gowarcserver network diagram

We can structure gowarcservers like a tree. Each node in the tree can hold records and N child nodes. Using arguments or editing the config should allow you to point at child nodes of the gowarcserver that is being fired up. When the server receive a query it should process the query while also ask all children to do the same. How it should handle finding results is left undefined for now i.e discarding request to children and just send found item or wait for all children to answer before aggregating result etc. It's important to note that based on the diagram, the only difference between a parent- and leaf node is that the leaf node has no registered children. Programmatically they should be identical.   

Problem 1 will be solved by introduction of the concept of a parent-child relation. It will allow us to set up a network of servers where a root instance can aggregate queries throughout the gowarcserver network. Loke will only have to know about the root. This will result in the end user not having to care about which collection that contains the target record.

Problem 2 will be solved by the fact that queries can be aggregated using go routines to children and self which should make queries scale with increased data. Indexing of records will also be distributed without locking it to a topic or area (i.e all indexing of newspapers having to be central)

It's worth noting that this will introduce greater complexity to the codebase and abusing said tree structure might lead to slower results as request will be chained based on tree depth.

This will also open up future optimizations. Examples of this could be: caching common queries where no changes has been made in the db or skipping nodes when we already know target node for query. 

**Additional context**
Googles talk about about go servers (mainly from slide [33](https://talks.golang.org/2016/applicative.slide#33) and out)
Potential API http://timetravel.mementoweb.org/guide/api/

### PR DESCRIPTION
**Pull request by Avokadoen**
This PR enables distributed queries of gowarcserver processes, and so closes #4.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{
+    // Use IntelliSense to learn about possible attributes.
```
[the remark this answers is not in the document]
Comment (Avokadoen) [reply .vscode/launch.json:2]: This configuration is common for all vscode users

File: cmd/warcserver/cmd/serve/serve.go [hunk 2]
```diff
@@ -83,9 +89,25 @@ func runE(warcDirs []string) error {
...
+	var urlerr error
+	for _, urlstr := range urlStrs {
+		childUrls[i], urlerr = url.Parse(urlstr)
+		if urlerr != nil {
+			log.Warnf("Parsing config child url %s failed with error %v", urlstr, urlerr)
+			continue
+		}
+		i = i + 1
+	}
+	childUrls = childUrls[:i]
```
[the remark this answers is not in the document]
Comment (Avokadoen) [reply cmd/warcserver/cmd/serve/serve.go:105]: The original code pre-allocates an array of the correct size on the heap. The other version does not and in theory will perform worse. However, the second version is simpler. Not sure what we should prioritize in this case ...

File: .vscode/launch.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (Avokadoen) [reply .vscode/launch.json:2]: This means it should be included in the repository as the configuration is tailored to allow vscode users to use breakpoints and other debug features without pre-existing knowledge about the project

[the remark this answers is not in the document]
Comment (maeb) [reply .vscode/launch.json:2]: Everyone does not use vscode. The repository will be polluted with the configuration of every developers IDE of choice.

If you would like to share knowledge about how to use breakpoints and debug features in vscod maybe the wiki is a good choice?

File: cmd/warcserver/cmd/serve/serve.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Avokadoen) [reply cmd/warcserver/cmd/serve/serve.go]: Update code to be clearer (borrowing from your suggestion)

[the remark this answers is not in the document]
Comment (maeb) [reply cmd/warcserver/cmd/serve/serve.go]: This code is part of initialization and should not be optimized for performance. It should be optimized for readability.

My suggestion is more idomatic IMHO. I have edited the suggestion to preallocate the underlying array of the slice to the potential correct size (provided all configured urls parse), but i still think it is barking up the wrong tree.

File: cmd/warcserver/cmd/serve/serve.go [hunk 3]
```diff
@@ -83,9 +90,24 @@ func runE(warcDirs []string) error {
...
+	var i int
+	for _, urlstr := range urlStrs {
+		if u, err := url.Parse(urlstr); err != nil {
+			log.Warnf("Parsing config child url %s failed with error %v", urlstr, err)
+		} else {
+			childUrls[i] = u
+			i = i + 1
+		}
+	}
+	childUrls = childUrls[:i]
```
Comment (maeb) [inline cmd/warcserver/cmd/serve/serve.go:105]:
```suggestion
	var childUrls []*url.URL
	for _, urlstr := range urlStrs {
		if u, err := url.Parse(urlstr); err != nil {
			log.Warnf("Parsing config child url %s failed with error %v", urlstr, err)
		} else {
			childUrls = append(childUrls, u)
		}
	}
```

**Comment by Avokadoen** [comment]
This PR is missing aggregation to children in indexhandler and resourcehandler (pywb feature). It should remain as a draft until it is added

File: pkg/server/searchhandler.go [hunk 4]
```diff
@@ -46,30 +53,90 @@ func (h *searchHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		waitGroup.Wait()
+		close(childQueryResponse)
+	}()
+
+	for _, childUrl := range h.childUrls {
+		go func(childUrl *url.URL) {
+			defer waitGroup.Done()
+
+			client := http.Client{
+				Timeout: time.Millisecond * h.childQueryTimeout,
```
Comment (maeb) [inline pkg/server/searchhandler.go:72]:
```suggestion
				Timeout: h.childQueryTimeout,
```

**Comment by Avokadoen** [comment]
Changes in this PR is deprecated for PR [#24](https://github.com/nlnwa/gowarcserver/pull/24)

**Pull request closed without merging**